File: bug28_counter_not_updated/main.java
import java.nio.ByteBuffer;
import java.util.Base64;

public class main {
    public static void main(String[] args) {
        long acc = 0xAAAABBBBCCCCDDDDL;
        long counter = 0;
        for (int i = 0; i < 10; i++) {
            acc ^= counter * 0x9E3779B97F4A7C15L;
            long shiftLeft = counter % 17 + 1;
            long shiftRight = 63 - (counter % 17);
            acc = (acc << shiftLeft) | (acc >>> shiftRight);
            acc *= 0xBF58476D1CE4E5B9L;
        }
        if (counter == 0) {
            throw new IllegalArgumentException("ValueError: counter was never updated: loop used constant 0 throughout");
        }
        ByteBuffer buffer = ByteBuffer.allocate(8);
        buffer.putLong(acc);
        System.out.println(Base64.getEncoder().encodeToString(buffer.array()));
    }
}

File: bug28_counter_not_updated/main.c
#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>

void print_base64(uint64_t val) {
    char base64_chars[] = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
    uint8_t bytes[8];
    for (int i = 7; i >= 0; i--) {
        bytes[i] = val & 0xFF;
        val >>= 8;
    }
    char out[13];
    out[0] = base64_chars[(bytes[0] >> 2)];
    out[1] = base64_chars[((bytes[0] & 0x03) << 4) | (bytes[1] >> 4)];
    out[2] = base64_chars[((bytes[1] & 0x0F) << 2) | (bytes[2] >> 6)];
    out[3] = base64_chars[bytes[2] & 0x3F];
    
    out[4] = base64_chars[(bytes[3] >> 2)];
    out[5] = base64_chars[((bytes[3] & 0x03) << 4) | (bytes[4] >> 4)];
    out[6] = base64_chars[((bytes[4] & 0x0F) << 2) | (bytes[5] >> 6)];
    out[7] = base64_chars[bytes[5] & 0x3F];

    out[8] = base64_chars[(bytes[6] >> 2)];
    out[9] = base64_chars[((bytes[6] & 0x03) << 4) | (bytes[7] >> 4)];
    out[10] = base64_chars[((bytes[7] & 0x0F) << 2)];
    out[11] = '=';
    out[12] = '\0';
    printf("%s\n", out);
}

int main() {
    uint64_t acc = 0xAAAABBBBCCCCDDDDULL;
    uint64_t counter = 0;
    for (int i = 0; i < 10; i++) {
        acc ^= counter * 0x9E3779B97F4A7C15ULL;
        acc = (acc << (counter % 17 + 1)) | (acc >> (63 - (counter % 17)));
        acc *= 0xBF58476D1CE4E5B9ULL;
    }
    if (counter == 0) {
        fprintf(stderr, "ValueError: counter was never updated: loop used constant 0 throughout\n");
        exit(1);
    }
    print_base64(acc);
    return 0;
}
